### PR TITLE
Api cleanup for Module and Type Reference directive resolution

### PR DESCRIPTION
**Pull request by sheetalkamat**
- Added new API for module resolution and type reference resolution that returns resolutions with failed lookups and also handles the resolution mode, This change was across all hosts (`CompilerHost`, `LanguageServiceHost` and `ProgramHost` which is used by `watch` and `tsc --build`. ce8da016f42d105e337dcb1fc3a6d20ab2732839 12a16ce4550f33856fe16135987d152fa0f1a71d
- Reusing module resolutions while constructing program reusing old program, now adds the diagnostics to file processing diagnostics so these diagnostics are not missed if program is reused. bbc4674d3ca6cc21c8cd4370de27d42bd989c14f 6a2ba6de8b2ae3350a58503e7c24777f85a9e15c
- Removed the ResolutionInfo that was added as stop gap since those APis are anyways deprecated 16846c806cb19857fb3860912e6801fbc3b49a2b

### CODE REVIEW

File: tests/baselines/reference/moduleResolution/baseUrl-without-path-mappings-or-rootDirs.js [hunk 1]
```diff
@@ -13,10 +13,7 @@ Resolution:: {
     "resolvedFileName": "/root/folder2/file2.ts",
     "extension": ".ts",
     "isExternalLibraryImport": false
-  },
-  "failedLookupLocations": [],
-  "affectingLocations": [],
-  "resolutionDiagnostics": []
```
Comment (andrewbranch) [inline tests/baselines/reference/moduleResolution/baseUrl-without-path-mappings-or-rootDirs.js:19]: I always assumed these were non-optional for V8 monomorphism reasons (though I would expect `undefined` instead of an empty array might be the best for performance?). @DanielRosenwasser @rbuckton is there a guiding principle for how to approach these things? Does it even make sense to try to make optimizations without looking at data from deoptigate?

Comment (sheetalkamat) [reply tests/baselines/reference/moduleResolution/baseUrl-without-path-mappings-or-rootDirs.js:19]: Made these optional because we want to handle optional undefined. We can always set these to empty array like before but thought undefined instead of empty array would better (since it is mutable empty array so always new array)

Comment (andrewbranch) [reply tests/baselines/reference/moduleResolution/baseUrl-without-path-mappings-or-rootDirs.js:19]: Yeah, I’m sure the new arrays is the worst of all the options. There might be some benefit to initializing to `undefined` but I’m not sure.

File: src/compiler/types.ts [hunk 2]
```diff
@@ -7337,15 +7335,34 @@ export interface CompilerHost extends ModuleResolutionHost {
...
+    resolveTypeReferenceDirectives?(typeReferenceDirectiveNames: string[] | readonly FileReference[], containingFile: string, redirectedReference: ResolvedProjectReference | undefined, options: CompilerOptions, containingFileMode?: ResolutionMode, resolutionInfo?: TypeReferenceDirectiveResolutionInfo): (ResolvedTypeReferenceDirective | undefined)[];
+    resolveModuleNameLiterals?(
+        moduleLiterals: readonly StringLiteralLike[],
+        containingFile: string,
+        redirectedReference: ResolvedProjectReference | undefined,
+        options: CompilerOptions,
+        containingSourceFile: SourceFile,
+        reusedNames: readonly StringLiteralLike[] | undefined,
+    ): readonly ResolvedModuleWithFailedLookupLocations[];
+    resolveTypeReferenceDirectiveReferences?<T extends FileReference | string>(
```
Comment (andrewbranch) [inline src/compiler/types.ts:7360]: This name is a lot 😵‍💫

File: src/compiler/program.ts [hunk 3]
```diff
@@ -4549,8 +4564,8 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
         if (!symlinks) {
             symlinks = createSymlinkCache(currentDirectory, getCanonicalFileName);
         }
-        if (files && resolvedTypeReferenceDirectives && !symlinks.hasProcessedResolutions()) {
-            symlinks.setSymlinksFromResolutions(files, resolvedTypeReferenceDirectives);
+        if (files && automaticTypeDirectiveResolutions && !symlinks.hasProcessedResolutions()) {
+            symlinks.setSymlinksFromResolutions(files, automaticTypeDirectiveResolutions);
```
Comment (andrewbranch) [inline src/compiler/program.ts:4589]: Why this change? If someone sets an empty `types` array but uses a bunch of triple slash type reference directives that resolve into symlinked node_modules, we’ll miss those symlinks.

File: src/compiler/types.ts (hunk 2 above)
Comment (sheetalkamat) [reply src/compiler/types.ts:7360]: open to suggestions

File: src/compiler/program.ts (hunk 3 above)
Comment (sheetalkamat) [reply src/compiler/program.ts:4589]: we are iterating over file.resolvedTypeReferenceDirectives so that should get covered?

**Review by andrewbranch — APPROVED** [review]
Looks good to me overall, but I think others should weigh in on the API changes. I also suspect we have some other infrastructure that might use `sourceFile.resolvedModules` even though it’s internal... @sandersn and/or I should check DefinitelyTyped-tools.

**Review by weswigham — APPROVED** [review]
Better unifying the mechanisms for type reference directive loading and import loading is 👍

File: src/compiler/types.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/types.ts:7360]: It at least makes sense, despite the repetition - you need actual references to type reference directives now, rather than `string`s (ideally). Related: Like the module name version, should we split the signature and attempt to deprecate the `string`-handling form?

Comment (andrewbranch) [reply src/compiler/types.ts:7360]: Yeah, the thing that makes me dislike the name is that it still takes a string, so seems like a bit of a misnomer.

File: src/services/types.ts [hunk 4]
```diff
@@ -357,9 +357,27 @@ export interface LanguageServiceHost extends GetEffectiveTypeRootsHost, MinimalR
      *
      * If this is implemented, `getResolvedModuleWithFailedLookupLocationsFromCache` should be too.
      */
-    resolveModuleNames?(moduleNames: string[], containingFile: string, reusedNames: string[] | undefined, redirectedReference: ResolvedProjectReference | undefined, options: CompilerOptions, containingSourceFile?: SourceFile, resolutionInfo?: ModuleResolutionInfo): (ResolvedModule | undefined)[];
+    /** @deprecated supply resolveModuleNameLiterals instead for resolution that can handle newer resolutions */
```
Comment (sandersn) [inline src/services/types.ts:360]: what does 'newer' mean? There might be a more descriptive word that would help an uninformed reader (like me)

**Review by sandersn — APPROVED** [review]
